### DISCUSSION TITLE
Simulation convergence problem

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/FEPX#59] [Q&A] [answered]
I'm working on a set of FEPX simulations with microstructures generated in Neper. I have not gotten any good results so I thought I'd look for some insights from the experts here.

The microstructure:
![2_10_3](https://user-images.githubusercontent.com/86078353/146817524-6864a193-0e59-4895-9464-0f500b1e80ff.png)
 This is an example of the microstructure I generated. It's a 3-scale tessellation by Neper, with the third scale being lamellar structures. There are two phases (HCP and BCC) assigned to the voronoi cells and lamellar cells, respectively. Each cell has its own orientations, and more specifically, voronoi cells are assigned random orientations while lamellar cells' orientations are calculated according to BOR. I'm also attaching the tessellation, mesh, simulation config files below.
[simulation_msh.txt](https://github.com/acmelab-ua/FEPX/files/7747416/simulation_msh.txt)
[tess.txt](https://github.com/acmelab-ua/FEPX/files/7747417/tess.txt)
[simulation_config.txt](https://github.com/acmelab-ua/FEPX/files/7747418/simulation_config.txt)

The mesh:
![1_10_3_msh](https://user-images.githubusercontent.com/86078353/146819349-898174c8-f25a-4c53-8278-24ce06d5723c.png)
The microstructures are meshed in Neper, with different sizes of mesh assigned to the voronoi cells and lamellar cells. I have tried to generate only a few thousand nodes (on a microstructure with ~10 cells) and ran it on my desktop (mpirun with 4 cores), and about 200k nodes (on microstructures with ~50 cells) and ran with clusters (64 cores). None of them converged (output attached below).
[output.txt](https://github.com/acmelab-ua/FEPX/files/7747401/output.txt)

My hypothesis is that, 
1. the mesh quality is poor. Although I have tried very fine meshes (200k for 50 cells, imo is already very fine) and got no luck. I have also set the mesh quality to the highest in Neper when doing the meshing.
2. The orientations are randomly assigned, which means their stiffness along loading direction are very different. This could be a reason why it is hard to converge?

I have successfully ran simulations with simpler setups so no doubt that fepx works properly.
Any input would be very much appreciated!!

**Comment by rquey** [comment, marked as answer]
There might be problems with both your Neper and FEPX inputs.

As for the mesh generated by Neper, it is not very good... You should make sure to use `-reg 1` and `-sel val` in `-T`, where `val` is a fraction of the lamella width (10%, typically), to remove the small edges. In `-M`,  using `-pl 100` would help, to avoid the mesh overrefinements. This should allow you to decrease the value of `-rcl`/`-cl` without increasing the number of elements compared to your current commands.

```
neper -T -n 10::3::from_morpho -morpho "::::lamellar(w=0.02:0.2)" -reg 1 -sel 0.002  -o d
neper -M d.tess -pl 1000 -cl 0.08 -order 2
neper -V d.msh -print d
```

![image](https://user-images.githubusercontent.com/2659376/146823472-2335916c-5d34-4a49-9c26-fdae054e1a04.png)

I suggest you try again to run your FEPX simulation with such a mesh.

**Reply by mkasemer** [reply]
I agree with the above comment. The mesh quality of your original mesh is quite poor, and these simulations (lamellar microstructures) are somewhat difficult to run, owing to the small features present in the tessellation. @rquey 's suggestions will reduce the small features, and facilitate a cleaner mesh. This may help.

**Reply by xshang93** [reply]
Great! Thank you so much for testing it out, @rquey . I will try this. But before I start,
1. I did try reg -1 before, and below are what I got with/without regularization turned on,
![1_20_3_reg](https://user-images.githubusercontent.com/86078353/146824481-a29fd6f8-9261-440b-a5db-b2d23540750d.png)
![1_20_3](https://user-images.githubusercontent.com/86078353/146824488-90236915-b7eb-4dad-8b4a-7f08bdbc98bc.png)
I have noticed that by turning on regularization, the results shows many slender lamellars with sharp tips (which can be seen in the pictures above). Plus, the one with reg turned on failed to be meshed while the other one succeed. That's why I ditched regularization in my current simulations;
2. I have not done much FEA for microstructures, but from my past experience with structural simulations, a rule of thumb is that you want at least two layers of meshes for a face. In the mesh you showed above, each lamellar has only one layer. This might have no effect on global properties but I also wanted to get the local stress/strain distributions, so this could be problematic? Any thoughts on this?

Thanks!

**Reply by xshang93** [reply]
> I agree with the above comment. The mesh quality of your original mesh is quite poor, and these simulations (lamellar microstructures) are somewhat difficult to run, owing to the small features present in the tessellation. @rquey 's suggestions will reduce the small features, and facilitate a cleaner mesh. This may help.

Thank you @mkasemer for your reply. I will try @rquey 's suggestions and get back with updates.

**Reply by mkasemer** [reply]
> 1. I did try reg -1 before, and below are what I got with/without regularization turned on,
>    I have noticed that by turning on regularization, the results shows many slender lamellars with sharp tips (which can be seen in the pictures above). Plus, the one with reg turned on failed to be meshed while the other one succeed. That's why I ditched regularization in my current simulations;

Regularization is necessary unless you are willing to refine your mesh to be such that there is not a high degree of selective refinement near small features. In general, this usually necessitates a very large mesh. The consequence of not considering regularization is that your mesh will have selective refinement and, likely, elements of poor aspect ratio, which will lead to premature simulation failure. In your case, it looks like the lamellae are rather thin to begin with, which may necessitate a large mesh.

The regularization options may be tuned to your lamellar width (10% of the width), such that you do not delete those edges, as @rquey suggests, and your lamella bounds will still be (roughly) parallel, rather than converging to a point.

> 2. I have not done much FEA for microstructures, but from my past experience with structural simulations, a rule of thumb is that you want at least two layers of meshes for a face. In the mesh you showed above, each lamellar has only one layer. This might have no effect on global properties but I also wanted to get the local stress/strain distributions, so this could be problematic? Any thoughts on this?

I agree that it is desirable to have multiple "layers" of elements through the thickness of the lamella. This was necessary to study the development of plasticity in a titanium microstructure in a previous study ([https://doi.org/10.1016/j.jmps.2017.03.013](https://doi.org/10.1016/j.jmps.2017.03.013)). However, this also then necessitated a rather large finite element mesh (2M+ elements), which ultimately required on the order of 32x8 processors to simulate.

**Reply by xshang93** [reply]
>
> > 2. I have not done much FEA for microstructures, but from my past experience with structural simulations, a rule of thumb is that you want at least two layers of meshes for a face. In the mesh you showed above, each lamellar has only one layer. This might have no effect on global properties but I also wanted to get the local stress/strain distributions, so this could be problematic? Any thoughts on this?
> 
> I agree that it is desirable to have multiple "layers" of elements through the thickness of the lamella. This was necessary to study the development of plasticity in a titanium microstructure in a previous study (https://doi.org/10.1016/j.jmps.2017.03.013). However, this also then necessitated a rather large finite element mesh (2M+ elements), which ultimately required on the order of 32x8 processors to simulate.

Right. I did notice how time consuming (owing to the large amount of meshes) the simulations were in this paper - In fact I relied heavily on this paper to set up my simulations. It's a great work!

**Reply by xshang93** [reply]
@rquey @mkasemer Thanks to your inputs, I got a much better mesh with much less meshes to solve.
![2_10_3_msh](https://user-images.githubusercontent.com/86078353/146882995-4b0f2b36-09e0-4e42-b7ad-9aabdd0912fb.png)
 I have even finished a simulation real fast!

However, as I was trying to post process it in Neper, new problem arose: `Error  : Different numbers of slip systems are not supported (complain!).` According to https://github.com/rquey/neper/discussions/163, neper does not support multi-phase results: is it still the case? I'm on v4.2.1-17. If this is the case, is there a workaround for HCP+BCC results?

EDIT: I just found from neper's manual that starting from v4.2.0, multiphase is supported by -S. So what am I encountering here?

**Reply by rquey** [reply]
If HCP and BCC have different numbers of slip systems, this is what the problem is.

**Reply by mkasemer** [reply]
Yes, this is the cause (18 and 12 slip systems). In this case, @xshang93, it is a current limitation with Neper to use -S module.

**Reply by xshang93** [reply]
@rquey @mkasemer Do you have any suggestions in this case how I can post-process the results? More specifically, the local stress/strain distribution and global responses. I might have an idea for the global strain/stress curve, as I think it's just a matter of plotting the reaction force vs the strain. For the local strain/stress, however, I'm absolutely new to it. Any suggestion would be appreciated!

**Reply by rquey** [reply]
We are just starting with -S... Let me a few days to add full support of multiphase simulations.

**Reply by xshang93** [reply]
Sounds great. Thank you for looking into this. Let me know if there is anything I can do to help - If I need to figure this out anyways, I think the best way is to contribute back to the community.

**Reply by xshang93** [reply]
Hi @rquey ! I wonder if there are any updates on the support for multi-phase? I don't mean to push you of course, just checking.

Thank you for all the help and support!

**Reply by rquey** [reply]
> Hi @rquey ! I wonder if there are any updates on the support for multi-phase? I don't mean to push you of course, just checking.
> 
> Thank you for all the help and support!

Sorry for the delay. I've made a preliminary version (on the Neper repo) that should allow you to process your simulation (given your `simulation.config` file, and as long as you do not process `slip` or `sliprate` variables). Version 4.2.1-26. I'm working on a complete fix.

For more discussions on this Neper-specific issue, please open an issue on the Neper repo.

> Sounds great. Thank you for looking into this. Let me know if there is anything I can do to help - If I need to figure this out anyways, I think the best way is to contribute back to the community.

I like that. If you'd like to contribute to Neper, there is a possibility here: https://github.com/rquey/neper/issues?q=is%3Aissue+is%3Aopen+label%3Abeginner.

**Reply by xshang93** [reply]
> Sorry for the delay. I've made a preliminary version (on the Neper repo) that should allow you to process your simulation (given your simulation.config file, and as long as you do not process slip or sliprate variables). Version 4.2.1-26. I'm working on a complete fix.

That's awesome. Thank you for working on this. I will test it out later.

> For more discussions on this Neper-specific issue, please open an issue on the Neper repo.

Sure thing. I will open up an issue for discussions on this matter.

> I like that. If you'd like to contribute to Neper, there is a possibility here: https://github.com/rquey/neper/issues?q=is%3Aissue+is%3Aopen+label%3Abeginner.

I'd like to look into this. Allow me a few days to report back.
